### PR TITLE
cryptsetup: add a timeout for waiting for FIDO2/PKCS#11/TPM2 devices

### PR DESCRIPTION
**Pull request by poettering**
Fixes: #19739

### CODE REVIEW

**Comment by bluca** [comment]
> Fixes: #10739

Wrong issue number I imagine?

File: src/cryptsetup/cryptsetup.c [hunk 1]
```diff
@@ -738,7 +739,16 @@ static int make_security_device_monitor(sd_event *event, sd_device_monitor **ret
 
 static bool libcryptsetup_plugins_support(void) {
 #if HAVE_LIBCRYPTSETUP_PLUGINS
-        return crypt_token_external_path() != NULL;
+        int r;
+
+        /* Permit a way to disable libcryptsetup token module support, for debugging purposes. */
+        r = getenv_bool("SYSTEMD_CRYPTSETUP_USE_TOKEN_MODULE");
```
Comment (bluca) [inline src/cryptsetup/cryptsetup.c:807]: Add to docs?

**Comment by yuwata** [comment]
LGTM. But Ubuntu CIs are red.

**Comment by bluca** [comment]
> LGTM. But Ubuntu CIs are red.

There are testbed issues, so we can ignore those. But the new env var could use being documented.

**Comment by oniko** [comment]
> (BTW, @oniko: i am wondering, what's the rationale for going through the token plugins from systemd-cryptsetup? I mean, we implement the non-plugin codepaths anyway, hence what do we gain in systemd-cryptsetup to go via the plugins for that? I am tempted to drop those codepaths, and not bother for the token modules that are anyway implemented in systemd code.

Just to reassure myself. You'd want to drop crypt_activate_by_token* libcryptsetup API usage in systemd-cryptsetup completely?

If so, please consider not doing so. I'll try to explain below:

> My reason for that: I'd like to rework the FIDO2 stuff so that we can ask the same question to multiple enrolled tokens, and then depending on the answer we get check if any of the enrolled ones match. But that means we want to reuse data we acquired from one token possibly on the next token and so on.

Can't you create custom context and pass it via usrptr in crypt_activate_by_token* calls? The internal logic in crypt_activate_by_token* guarantees that if you pass "systemd-fido2" token type as parameter the context (usrptr) gets passed only to systemd-fido2 tokens and you may keep that data there.

> that is easy if we don't bother with the plugins for the fido2 stuff. if we always go via the plugins also for the fido2 stuff passing this data is not as easy. hence my question: can we drop the plugin codepaths from systemd-cryptsetup? what would we lose?)

Frankly? Systemd itself perhaps not so much. But my goal was eliminate as much as possible code duplication in between systemd and cryptsetup. Recall that cryptsetup had it's own tpm2 integration on the table at the moment systemd implemented systemd-tpm2 token and associated logic. And we dropped our implementation in favor of systemd one, so that there was no userbase split.

There are other libcryptsetup internal  things. Libcryptsetup API calls respect keyslot priorities, it can easily see keyslot elligibility for current effective volume key (so you don't waste, just an example, PIN try on a token with unusable keyslot for whatever reason). Then there's reencryption (multiple effective volume keys) and so on.

So, is there anything we can do (on libcryptsetup side, do you miss any particular API) to avoid this? I guess the moment you drop crypt_activate_token* logic from systemd we will (sooner or later) end with systemd tokens being unusable from cryptsetup cli due to some future incompatibility.

**Comment by mbroz** [comment]
@poettering Could you please repond to @oniko comments?

**Pull request merged**